File: Java/timing_leak.java
public class TimingLeakExample {
    public static boolean insecureCheck(String input, String correct) {
        // Bad case: Branching based on secret value
        for (int i = 0; i < correct.length(); i++) {
            if (input.charAt(i) != correct.charAt(i)) {
                return false; // Timing leak occurs here
            }
        }
        return true;
    }

    public static boolean secureCheck(String input, String correct) {
        // Good case: Constant-time comparison (prevents timing attacks)
        int diff = 0;
        for (int i = 0; i < correct.length(); i++) {
            diff |= input.charAt(i) ^ correct.charAt(i);
        }
        return diff == 0;
    }

    public static void main(String[] args) {
        String correctPassword = "secure123";
        String input = "secure123";

        long start = System.nanoTime();
        boolean result = insecureCheck(input, correctPassword);
        long end = System.nanoTime();
        System.out.println("Insecure Check Time: " + (end - start) + " ns, Result: " + result);

        start = System.nanoTime();
        result = secureCheck(input, correctPassword);
        end = System.nanoTime();
        System.out.println("Secure Check Time: " + (end - start) + " ns, Result: " + result);
    }
}

File: C/timing_leak.c
#include <stdio.h>
#include <string.h>
#include <time.h>

// Insecure function (leaks timing information)
int insecureCheck(const char *input, const char *correct) {
    for (size_t i = 0; i < strlen(correct); i++) {
        if (input[i] != correct[i]) {
            return 0; // Timing leak occurs here
        }
    }
    return 1;
}

// Secure function (constant-time comparison)
int secureCheck(const char *input, const char *correct) {
    int diff = 0;
    size_t len = strlen(correct);
    
    for (size_t i = 0; i < len; i++) {
        diff |= input[i] ^ correct[i];  // No branching
    }
    
    return diff == 0;
}

int main() {
    const char *correctPassword = "secure123";
    const char *input = "secure123";

    // Measure time for insecure check
    clock_t start = clock();
    int result = insecureCheck(input, correctPassword);
    clock_t end = clock();
    printf("Insecure Check Time: %lu ns, Result: %s\\n", (end - start), result ? "Match" : "No Match");

    // Measure time for secure check
    start = clock();
    result = secureCheck(input, correctPassword);
    end = clock();
    printf("Secure Check Time: %lu ns, Result: %s\\n", (end - start), result ? "Match" : "No Match");

    return 0;
}
